Add DrawTarget adapters

rfuest opened the pull request:
…traits

Hi! Thank you for helping out with Embedded Graphics development! Please:

- [ ] Check that you've added passing tests and documentation
- [ ] Add a simulator example(s) where applicable
- [ ] Add a `CHANGELOG.md` entry in the **Unreleased** section under the appropriate heading (**Added**, **Fixed**, etc) and appropriate crate (`embedded-graphics`, `simulator`, `tinytga`, `tinybmp`) if your changes affect the **public API**
- [ ] Run `rustfmt` on the project
- [ ] Run `./build.sh` (Linux/macOS only) and make sure it passes. If you use Windows, check that CI passes once you've opened the PR.

## PR description

I've started to experiment with the `TranslatedDrawTarget` (see #355). The `primitives-fill.rs` example demonstrates how it can be used. To help with the implementation I've also implemented `TranslatedPixelIterator`. Both objects can be created using the `translate` method provided by extension traits for `DrawTarget` and `Iterator<Item = Pixel<C>>`.

One feature I'm not sure about is `drawable.draw_at(offset, &mut target)`, which is a shorthand for `drawable.draw(&mut target.translate(offset))`. This could be used to replace `Image::offset` and `Polyline::translate`, but maybe it's better to keep the offset in the drawable itself.

The PR still needs some polish, but I would appreciate some initial feedback.

jamwaffles commented:
Didn't we decide to call it `SubDrawTarget`? I had it in my head that it could be a different size from the parent `DrawTarget` instead of just an offset of it.

In terms of API, what about flipping it on its head somewhat (excuse the names):

```rust
// Get a display instance that impls DrawTarget
let display = Disp::new();

// Sub draw target with an offset of (10, 20) and size (30, 40)
let sub = display.sub_draw_target(Rectangle::new(Point::new(10, 20), Size::new(30, 40)));

let circle = Circle::new(Point::zero(), 10).into_styled(...);

// Draw at display origin
circle.draw(&mut display).unwrap();

// Draw at subtarget origin
circle.draw(&mut sub).unwrap();
```

This would allow the removal of `DrawableExt` and keep the same API for `DrawTarget` and `SubDrawTarget`.

Here's a pile of questions:

Should we mirror (or implement) the `Transform` API for `SubDrawTarget` and use `.translate()` and `.translate_mut()`? This would allow sub draw targets to be moved around.

Do we want to keep the shorthand top level `display.translate()`, or keep the API simpler with a single method to get a sized sub draw target. `translate` as a name is an odd word when called on a top level display IMO - it feels like it should move the simulator window/physical display around which doesn't make sense for obvious reasons.

Can sub draw targets be larger than their parent? IMO this should be possible. If not, should we convey creation failure with a `Result`?

Can the subtarget be resized? If so, do we want to make a variant that can't which is just an offset of the parent DrawTarget? I'm not sure this really gives any benefit and just complicates the API so I'd say we just have one that can be resized for now, IMO.

We also might want to allow configuration of subtarget overdraw - some applications may want clipping, others not.

It might help to describe the use cases you have in mind for this API so we can align on a final goal for it. The 2 that come to mind for me are:

1. Supporting that "footer" on the ST7789 that was discussed a while ago
2. Features to support layout/windowing engines for something like an oscilloscope interface or embedded-layout.

By the way I like the recursive nature of this feature - I think that could be really useful for more complex layouts/UI.

bugadani commented:
In general, I don't think resize/extend is a good idea. Even though the TDT is not a clipping area, if there will ever be a clipping implementation, the API will probably be similar to not confuse the users.

The problem is, if something gets a mutable reference, if they can resize it or otherwise modify, they can essentially break out of the clipping rectangle. Even worse (granted, it's subjective), they can in principle assign a different area to the reference and thus modify the area the caller works with. If there can ever exist a situation where the callee is not completely trusted, it can cause weird issues IMO. Maybe there are ways to prevent these kinds of code to compile, though, I'm not fluent enough in Rust to say.

Speaking of confusion, although it's not mean to compile, @jamwaffles's code can't compile because it violates Rust's aliasing rules: as long as the sub drawtarget lives, the main draw target can't be referenced. Assuming, of course, that the sub target holds a mutable reference to the parent draw target.

Just my .02€, feel free to disregard

rfuest commented:
> Didn't we decide to call it `SubDrawTarget`? I had it in my head that it could be a different size from the parent `DrawTarget` instead of just an offset of it.

Because this is only implements a part of the functionality of `SubDrawTarget` I decided to call it `TranslatedDrawTarget` instead. IMO it's better to keep translation and clipping separate, because there are use case where you only need one of the two. (see https://github.com/jamwaffles/embedded-graphics/issues/355#issuecomment-639554399)

> In terms of API, what about flipping it on its head somewhat (excuse the names): [... code ...]

As @bugadani mentioned your code example wouldn't work, but when you reorder the instructions it will already work this way (without the clipping):

```rust
let display = Disp::new();

let circle = Circle::new(Point::zero(), 10).into_styled(...);
circle.draw(&mut display).unwrap();

let mut sub = display.translate(Point::new(10, 20));
circle.draw(&mut sub).unwrap();
```
This is used in the `primitives-fill.rs` example.

> This would allow the removal of DrawableExt and keep the same API for DrawTarget and SubDrawTarget.

`DrawableExt` wasn't intended to be a replacement for the code in your example, but an alternative implementation of the offset function in `Image` and `Polyline`. But I think I'll remove it from this PR to keep it focused.

> Should we mirror (or implement) the `Transform` API for `SubDrawTarget` and use `.translate()` and `.translate_mut()`? This would allow sub draw targets to be moved around.

I consider `TranslatedDrawTarget` to be a short lived object that gets created while something is drawn and which isn't preserved across multiple draw calls. Because of this I don't think it's necessary to mutate it.

> Do we want to keep the shorthand top level `display.translate()`, or keep the API simpler with a single method to get a sized sub draw target. `translate` as a name is an odd word when called on a top level display IMO - it feels like it should move the simulator window/physical display around which doesn't make sense for obvious reasons.

I would like to keep the functionality, but I'm open for suggestions on better names.

> Can sub draw targets be larger than their parent? IMO this should be possible. If not, should we convey creation failure with a `Result`?
> Can the subtarget be resized? If so, do we want to make a variant that can't which is just an offset of the parent DrawTarget? I'm not sure this really gives any benefit and just complicates the API so I'd say we just have one that can be resized for now, IMO.

Interesting questions, but if this PR only adds translation they aren't not relevant here. Can you move this question to #355?

> We also might want to allow configuration of subtarget overdraw - some applications may want clipping, others not.

This would be solved by splitting `SubDrawTarget` into `TranslatedDrawTarget` and `ClippingDrawTarget`.

> Supporting that "footer" on the ST7789 that was discussed a while ago

I didn't notice this discussion, can you provide a link to it?

jamwaffles commented:
Thanks both for educating me. I should've dug a bit deeper into my suggested code.

@rfuest All those points SGTM so I won't respond to each individually other than:

> I didn't notice this discussion, can you provide a link to it?

This happened in Matrix a long time ago IIRC, I don't think I can find a link back to it unfortunately.

jamwaffles commented:
Thanks both for educating me. I should've dug a bit deeper into my suggested code.

@rfuest All those points SGTM so I won't respond to each individually other than:

> I didn't notice this discussion, can you provide a link to it?

This happened in Matrix a long time ago IIRC, I don't think I can find a link back to it unfortunately.

> One feature I'm not sure about is drawable.draw_at(offset, &mut target)

I'd prefer to leave this out, at least for now. If there's a pressing need for it in the future, we can always add it back in.

rfuest commented:
`translate` is a common name in canvas APIs: https://www.cairographics.org/manual/cairo-Transformations.html#cairo-translate and https://www.cairographics.org/manual/cairo-Transformations.html#cairo-translate

But because we return an object instead of mutating a context we could use a different name. I would be OK with `translated`.

bugadani commented:
> But because we return an object instead of mutating a context we could use a different name. I would be OK with `translated`.

I think the Rust API guidelines say that mutating versions of a certain API should be named `*_mut`. Just based on this I think `translate` is fine

jamwaffles commented:
`translated` it is then :)

rfuest commented:
I've renamed `translate` to `translated` and removed the extension trait for `Drawable`s.

rfuest commented:
I've made the `draw_target` module public, to have a place to put the new types into. In #330 we moved `DrawTarget` to the root of the crate, but IMO it would now be better to remove it from the crate root and only export it in the `draw_target` module, because it breaks the doc links otherwise. And I think that having the trait exported in one place and the prelude should be enough. This is a breaking change, but it won't affect users that have imported `DrawTarget` via the prelude.

jamwaffles commented:
That's fine. It _is_ breaking but it's not a difficult fix to make. Can you add a breaking changelog entry just to help anybody that might get stuck.

rfuest commented:
> Can you add a breaking changelog entry just to help anybody that might get stuck.

Sure, I'll include it when I write the changelog entries for the finished PR.

rfuest commented:
I've pushed some more changes to this PR, like adding a first version of `CroppedDrawTarget`.

Having `DrawTarget::size()` didn't make sense anymore, because the origin of the draw target's bounding box might not be at `(0, 0)`. Instead `DrawTarget`s are now required to implement `Dimensions`.

`OriginDimensions` was added to make implementing `Dimensions` a little easier and to provide a type level way to determine if `bounding_box` will return a rectangle with `top_left == Point::zero()`, which could be useful as a replacement for `ImageDimensions`. Is the name for this trait OK or should we change it?

bugadani commented:
Maybe my problem comes from my inexperience with drawing apis, but for me, this feels a bit weird. I was always under the assumption that cropping implies clipping. I'm also not sure why a cropped draw target has a size. If it doesn't do clipping, shouldn't it's size be infinite? (But in that case, what would it's position represent?)

rfuest commented:
If a drawing operation uses the `bounding_box` returned by the draw target for some drawing operations, `CroppedDrawTarget` can be used to pass a smaller portion of the base target to that drawing operation. It doesn't imply clipping because there are use cases that benefit from being able to draw outside the actual target. One example would be to implement drawing of shadows similar to `box-shadow` in CSS, where the shadow is drawn outside the element.

>  If it doesn't do clipping, shouldn't it's size be infinite? (But in that case, what would it's position represent?)

The `bounding_box` implementation was broken and should always return `top_left == Point::zero()` which might have caused some confusion.

jamwaffles commented:
> ... which could be useful as a replacement for ImageDimensions

Very nice. Will you remove `ImageDimensions` in this PR? Either way, it would close #367 which is a nice bonus.

> Is the name for this trait OK or should we change it?

I can't think of a better one, it's descriptive, and most people will include it through the prelude or not at all, so will rarely see the name. I think `OriginDimensions` is fine :)

rfuest commented:
> Will you remove `ImageDimensions` in this PR?

No, I will include that change in the `SubImage` PR.

bugadani commented on embedded-graphics/src/draw_target/clipped.rs:
```diff
@@ -6,6 +6,12 @@ use crate::{
+/// Created by calling [`clipped`] on any [`DrawTarget`].
+/// See the [`clipped`] method documentation for more.
+///
+/// [`DrawTarget`]: trait.DrawTarget.html
```
Just making a note here since I was once asked to revert intra-crate links to regular ones, because the feature is not yet stable. Right now it's targeted at 1.47.0, which is still some time away. To be fair, with embedded-layout, I'm also using these with a note that although the crate's MSRV is 1.40 or so, the docs require nightly to build.

rfuest commented:
I think this should be ready for review now. The implementation of `Clipped::fill_contiguous` still contains a TODO, but I would prefer to fix that in the `SubImage` PR.

jamwaffles commented on embedded-graphics/src/draw_target/mod.rs:
```diff
@@ -439,6 +447,161 @@ pub trait DrawTarget {
+    /// Its coordinate system is shifted so that the origin coincides with `area.top_left` in the
+    /// base draw target's coordinate system.
+    ///
+    /// The bounding box of the returned target will always be contained inside the bounding box
+    /// of the base target. If the requested `area` is overlapping the base target's bounding box
+    /// the intersection of the base target's bounding box and `area` will be used.
```
```suggestion
    /// The bounding box of the returned target will always be contained inside the bounding box
    /// of the base target. If any of the requested `area` lies outside the base target's bounding 
    /// box the intersection of the base target's bounding box and `area` will be used.
```

rfuest replied:
Thanks for the suggestion, I was struggling to get this paragraph right.

jamwaffles commented on embedded-graphics/src/draw_target/mod.rs:
```diff
@@ -439,6 +447,161 @@ pub trait DrawTarget {
+    /// The coordinate system of the created draw target is equal to the base target's coordinate
+    /// system. All drawing operations outside the bounding box will be clipped.
+    ///
+    /// The bounding box of the returned target will always be contained inside the bounding box
+    /// of the base target. If the requested `area` is overlapping the base target's bounding box
+    /// the intersection of the base target's bounding box and `area` will be used.
```
```suggestion
    /// The bounding box of the returned target will always be contained inside the bounding box
    /// of the base target. If any of the requested `area` lies outside the base target's bounding 
    /// box the intersection of the base target's bounding box and `area` will be used.
```

rfuest commented on embedded-graphics/src/draw_target/clipped.rs:
```diff
@@ -0,0 +1,222 @@
+/// [`DrawTarget`]: trait.DrawTarget.html
+/// [`clipped`]: trait.DrawTargetExt.html#tymethod.clipped
+#[derive(Debug)]
+pub struct Clipped<'a, T>
```
I tried to follow the same naming convention that is used for iterator adapters: https://rust-lang.github.io/api-guidelines/naming.html#iterator-type-names-match-the-methods-that-produce-them-c-iter-ty

I agree that `Clipped` on its own is a bit ambiguous, but the complete type produced by `clipped` is nicer without the additional `Target` IMO. For example: `Clipped<Ili9341<_, _>>` vs `ClippedTarget<Ili9341<_, _>>`

jamwaffles replied:
I didn't consider it with the extra types inside it. That does look a lot better.

rfuest commented on embedded-graphics/src/draw_target/cropped.rs:
```diff
@@ -0,0 +1,209 @@
+    T: DrawTarget,
+{
+    fn bounding_box(&self) -> Rectangle {
+        Rectangle::new(Point::zero(), self.size)
```
The returned `bounding_box` is using the coordinate system of the sub target not the parent one.

rfuest commented on embedded-graphics/src/draw_target/translated.rs:
```diff
@@ -0,0 +1,190 @@
+    T: DrawTarget,
+{
+    fn bounding_box(&self) -> Rectangle {
+        self.target.bounding_box().translate(-self.offset)
```
No it shouldn't. In the sub target coordinate system the bounding box of the parent will have `top_left` at a negative coordinate if `offset` is positive.

rfuest commented on embedded-graphics/src/draw_target/translated.rs:
```diff
@@ -0,0 +1,190 @@
+        let translated = display.translated(Point::new(1, 3));
+
+        assert_eq!(
+            display_bb.translate(-Point::new(1, 3)),
```
I didn't want to introduce a dependency on the `MockDisplay` size.

jamwaffles replied:
Ah ok, makes sense.

rfuest commented:
> My mental model moves the origin of each sub-target to that target's top-left corner, whilst I think yours might leave it at the top level "true" origin. Is that right?

Both `Cropped` and `Translated` shift the origin of drawing operations. But they differ in the rectangle returned by `bounding_box`. While `Cropped` returns a new bounding box with the origin at (0, 0), `Translated` returns a shifted version of the original bounding box.

`Cropped` doesn't shift the origin and only reports a different `bounding_box`.

> I think you mentioned using `Translated<Cropped>` or `Translated<Clipped>` if "windowing" behavior is required.

If the drawing operation can be trusted to not draw outside of the `bounding_box` `Cropped` can be used on its own. Otherwise `Clipped<Translated>` or `Translated<Clipped>` can be used.

> ...although an example of combining a couple of the different targets to e.g. produce a window with shifted origin would be useful to add to the `DrawTargetExt` documentation.

Yes that's a good idea. I think a simulator example should also be added. But can we add this in another PR? I can create an issue to track this.

jamwaffles commented:
Ok, thanks for explaining how this works. A compound example would've helped un-confuse me for sure ;) but yeah let's put that in a different PR. Please open an issue to track.

rfuest commented:
> Please open an issue to track.

Done.

jamwaffles commented:
Awesome, thanks. I've been reading through the docs a bit and thought `parent draw target` might be a better name than `base draw target`. Thoughts?

rfuest commented:
> Thoughts?

Good idea! I've changed the docs and also renamed the `target` field in the sub draw targets into `parent` to make the name more descriptive.
